Redesign crates, add Azure support

auguwu opened the pull request:
This crate does a whole lot of re-designing for the long run.

* Removed `remi_core` as `remi` should serve the same purpose
* Restructure arguments to methods that rely on `String`s or `&[u8]` to use an impl of `Into<T>`.
* Having methods "panic" to seal a mut reference is stupid and I wish I never did this! This uses `mut self` instead of `&mut self`.

This crate also adds experimental support for [Azure Blob Storage](https://azure.microsoft.com/en-us/products/storage/blobs) via [`azure_blob_storage`](https://github.com/Azure/azure-sdk-for-rust) crate and [Google Cloud Storage](https://cloud.google.com/storage) via [`google-cloud-storage`](https://github.com/yoshidan/google-cloud-rust/tree/main/storage) crate.

auguwu committed:
redesign crates API (not finished), decommission `remi_core`, add Nix-related files

auguwu committed:
fix merge conflicts from master

auguwu committed:
Add @spotlightishere as a code owner

auguwu committed:
fix errors with S3 impl, finish code for all of that

auguwu committed:
update license to 2024, other fixes

auguwu committed:
fix merge conflict from `master` ~> `feat/crate-redesign`

renovate[bot] committed:
Update aws-sdk-rust monorepo

auguwu committed:
add gridfs impl. again with better code

auguwu committed:
fix clippy warnings

auguwu committed:
fix even more clippy warnings

auguwu committed:
add `remi-azure` code (not finished)

auguwu committed:
[ci] run tests

auguwu committed:
fix `dbg_config` test

auguwu committed:
finish Azure implementation

auguwu committed:
disable doctests in src/options.rs for now

auguwu committed:
ignore remi doctests, use `trace` and `warn` macros in remi_fs::StorageService::normalize

auguwu committed:
update README

auguwu committed:
[fs] add example

auguwu committed:
add example notice in README

auguwu committed:
Merge branch 'master' of github.com:Noelware/remi-rs into feat/crate-redesign

auguwu committed:
[remi-azure] Disable Docker-related tests on Windows

spotlightishere commented on crates/azure/src/service.rs:
```diff
@@ -0,0 +1,541 @@
+        // in CI (GitHub Actions), it will pull the `azurite` image as a Windows container and Microsoft
+        // doesn't ship Windows containers of the `azurite` image, so we just ignore the test alltogether
+        //
+        // TODO(@auguwu): how to fix this? :3
```
Would it be possible to split this out into a separate issue?

spotlightishere commented on crates/gridfs/src/service.rs:
```diff
@@ -0,0 +1,439 @@
+        stream.write_all(&options.data[..]).await?;
+        stream.close().await
+
+        // TODO(@auguwu): add metadata to document that was created and the given content type
```
Similarly, can this be split out into its own issue?

spotlightishere reviewed (approved):
This is an extremely large pull request, so I can't quite do an in-depth review! However, from looking at the general scope of changes, this seems like a promising direction - and certainly assists with moving a lot of the AWS S3 SDK specific things out!

I left only quick comments asking to please make `TODO`s in source also become GitHub issues - would appreciate your insight!

auguwu commented:
> I left only quick comments asking to please make `TODO`s in source also become GitHub issues - would appreciate your insight!

I'll make marked `TODO`s as issues once this PR is merged.

auguwu committed:
Remove `remi-gcs` crate
The library I wanted to use is not what I had planned in mind of designing the `remi-gcs` crate. There can be a
community maintained one or when people request Google Cloud Storage support, then we'll talk about it via issues.

auguwu commented:
I decided to not work on `remi-gcs`, as I'll plan to do it soon when there is an officially maintained Google Cloud SDK that is reliable or a community-maintained version can be present if wished.

auguwu committed:
remove `remi-gcs` reference from README
